pkg/projects/resource_project_role.go

```diff
@@ -0,0 +1,260 @@
+
+	// REST API returns all project roles, including ones with PREDEFINED type which can't be altered.
+	// We are only interested in the "CUSTOM" types that we can manipulate.
+	customRoles := roles[:0]
+	for _, role := range roles {
+		if role.Type == customRoleType {
+			customRoles = append(customRoles, role)
+		}
+	}
+
+	log.Printf("[TRACE] customRoles: %+v\n", customRoles)
+
+	return customRoles, nil
+}
+
+var updateRoles = func(projectKey string, terraformRoles []Role, m interface{}) ([]Role, error) {
+	log.Println("[DEBUG] updateRoles")
+	log.Printf("[TRACE] terraformRoles: %+v\n", terraformRoles)
+
+	projectRoles, err := readRoles(projectKey, m)
+	log.Printf("[TRACE] projectRoles: %+v\n", projectRoles)
+
+	rolesToBeAdded := difference(rolesToEquatables(terraformRoles), rolesToEquatables(projectRoles))
+	log.Printf("[TRACE] rolesToBeAdded: %+v\n", rolesToBeAdded)
+
+	rolesToBeUpdated := intersection(rolesToEquatables(terraformRoles), rolesToEquatables(projectRoles))
+	log.Printf("[TRACE] rolesToBeUpdated: %+v\n", rolesToBeUpdated)
+
+	rolesToBeDeleted := difference(rolesToEquatables(projectRoles), rolesToEquatables(terraformRoles))
+	log.Printf("[TRACE] rolesToBeDeleted: %+v\n", rolesToBeDeleted)
+
+	var errs []error
+
+	for _, role := range rolesToBeAdded {
+		log.Printf("[TRACE] %+v\n", role)
+		err := addRole(projectKey, role.(Role), m)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	for _, role := range rolesToBeUpdated {
+		log.Printf("[TRACE] %+v\n", role)
+		err := updateRole(projectKey, role.(Role), m)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	err = deleteRoles(projectKey, equatablesToRoles(rolesToBeDeleted), m)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		return nil, fmt.Errorf("failed to update roles for project: %s", errs)
+	}
+
+	return readRoles(projectKey, m)
+}
+
+var addRole = func(projectKey string, role Role, m interface{}) error {
+	log.Println("[DEBUG] addRole")
+
+	_, err := m.(*resty.Client).R().SetBody(role).Post(fmt.Sprintf(projectRolesUrl, projectKey))
+
+	return err
+}
+
+var updateRole = func(projectKey string, role Role, m interface{}) error {
+	log.Println("[DEBUG] updateRole")
+
+	_, err := m.(*resty.Client).R().SetBody(role).Put(fmt.Sprintf(projectRolesUrl, projectKey) + role.Name)
+
+	return err
+}
+
+var deleteRoles = func(projectKey string, roles []Role, m interface{}) error {
+	log.Println("[DEBUG] deleteRoles")
+
+	var errs []error
+	for _, role := range roles {
+		err := deleteRole(projectKey, role, m)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to delete roles from project: %v", errs)
+	}
+
+	return nil
+}
+
+var deleteRole = func(projectKey string, role Role, m interface{}) error {
+	log.Println("[DEBUG] deleteRole")
+	log.Printf("[TRACE] %+v\n", role)
+
+	_, err := m.(*resty.Client).R().Delete(fmt.Sprintf(projectRolesUrl, projectKey) + role.Name)
```

Comment:
If you're already using `Sprintf` then also using string concatenation is weird. Keep it consistent: `Sprintf` or `+`

Reply:
Good idea on using Resty way. Will update.